php/libraries/Candidate.class.inc

```diff
@@ -680,115 +678,258 @@ class Candidate
+            throw new LorisException(
+                'Invalid value passed for parameter "kind". Must be one of: ' .
+                '"PSCID", "ExternalID".'
+            );
+        }
+        // Query database for existing IDs to prevent duplicates.
+        $existingIDs = \Candidate::_getExistingIDs($kind, $prefix);
+
+        /* Read config settings from project/config.xml to retrieve the
+         * alphabet, length, and generation method (sequential or random) used
+         * to create new IDs.
+         */
+        $alphabet   = \Candidate::_getIDSetting($kind, 'alphabet');
+        $length     = \Candidate::_getIDSetting($kind, 'length');
+        $generation = \Candidate::_getIDSetting($kind, 'generation');
+
+        // Initialize minimum and maximum allowed values for IDs. Set the values
+        // to the lowest/highest character in $alphabet repeated $length times
+        // if the min or max is not configured in project/config.xml.
+        try {
+            $min = \Candidate::_getIDSetting($kind, 'min');
+        } catch (\LorisException $e) {
+            $min = str_repeat(strval($alphabet[0]), $length);
+        }
+        try {
+            $min = \Candidate::_getIDSetting($kind, 'max');
+        } catch (\LorisException $e) {
+            $max = str_repeat(strval($alphabet[count($alphabet) - 1]), $length);
+        }
+
+        // Check that the number of existing IDs does not exceed the range of
+        // possible values (given by the size of the alphabet to the power of
+        // number of characters in the ID).
+        if (count($existingIDs) >= (count($alphabet) ** $length)) {
+            throw new \LorisException(
+                'Cannot create new identifier because all valid identifiers ' .
+                'are in use!'
+            );
+        }
+
+        switch ($generation) {
+        case 'sequential':
+            /* If there are no existing IDs, create a new string
+             * with the first character of the alphabet repeated $length
+             * times.
+             */
+            if (count($existingIDs) < 1) {
+                $id = $min;
             } else {
-                $pscid = Candidate::generateRandomStringFromArrays($pscidArrays);
+                $id = \Candidate::_nextValueInSequence(
+                    $existingIDs,
+                    $length,
+                    $min,
+                    $max
+                );
             }
-
-            if ($pscid == '') {
-                throw new LorisException(
-                    "Well this is embarrassing, our automated PSCID generator "
-                    . "is about to enter an infinite loop."
+            break;
+        case 'random':
+            $id = \Candidate::_generateNewRandomID($alphabet, $length);
+            // If the random ID is already in use, increment it until we find
+            // an unused value.
+            if (in_array(
+                str_pad($id, $length, '0', STR_PAD_LEFT),
```

Comment:
```suggestion
                str_pad($id, $length, $alphabet[0], STR_PAD_LEFT),
```
as well as in line 813 and 906
might be alpha characters only

Reply:
Thanks, I'll make this change.

I think I'll also make the alpha characters uppercase instead of lowercase in the range.